### PR TITLE
docs(udon-sharp): complete initializer and synced-array edge guidance

### LINKED ISSUE #355: Task: Complete initializer and synced-array edge-case guidance for v4.0.0
## Purpose

Complete the two remaining operational edge cases related to #338 and #344 before the public v4.0.0 release. These notes are now release requirements under #349 rather than post-release follow-ups.

- Start Date: 2026-08-19
- Target Date: 2026-08-19
- Status: In Progress
- Shipping milestone: public v4.0.0
- Parent: #349
- Related reports: #338, #344

## Scope

1. Make the field-initializer consequence concrete: a nondeterministic call such as `Random.Range` is evaluated on the Editor side and its result is stored as a baked program default. It must not be presented as runtime randomness. Use `Start()` or lazy initialization only for local/per-client randomness; generate shared per-object or per-session state on the owner and synchronize it with `[UdonSynced]`.
2. Complete the synced-array revision guidance: a late joiner receives the latest revision in its first deserialization and would run an uninitialized one-shot guard once. Document how to establish the first received revision as a baseline when historical effects must not replay, while keeping durable state projection in the idempotent `ApplyValues()` path.
3. Update the always-loaded Skill/rules, detailed references, troubleshooting/cheatsheet surfaces where the behavior is routed.
4. Strengthen the existing #338 and #344 documentation contracts so both edge cases fail when removed or weakened.
5. Remove contradictory late-joiner recipes that call `RequestSerialization()` from `OnPlayerJoined` only to push unchanged state; late joiners already receive the latest synced values automatically. Apply this consistently across active references and distributed templates, while preserving join handlers that actually mutate authoritative state before serialization.

## Acceptance criteria

- [x] RED contract tests demonstrate that the current `dev` branch does not yet pin either edge case.
- [x] Field-initializer guidance explicitly distinguishes a baked `Random.Range` result from runtime randomness and gives correct alternatives for both local/per-client randomness and owner-generated shared state.
- [x] Synced-array revision guidance explicitly states the first-deserialization behavior for late joiners and gives a no-replay baseline pattern.
- [x] The late-joiner pattern does not suppress durable state/UI reconstruction and does not misrepresent the revision as ordering or stale-packet protection.
- [x] Active references and distributed templates do not require `RequestSerialization()` from `OnPlayerJoined` merely to resend unchanged state; legitimate join handlers that create or mutate authoritative state remain intact.
- [x] All relevant documentation contracts, hook suites, Skill validation, packaging checks, and `git diff --check` pass.
- [x] Independent Skill review finds no CRITICAL, HIGH, or MUST issues and confirms no part of #338/#344 was weakened.
- [x] A dev-targeted PR is opened with `Closes` for this task and `Refs #338, #344, #349`, then reaches CI/review completion before merge approval is requested.

## Out of scope

- The held `dev` to `main` release PR, tag, GitHub Release, and npm publish.
- Supporting SDK versions older than 3.10.4.
- Rewriting unrelated networking or initializer documentation.

### PR DESCRIPTION
**Pull request by niaka3dayo**
## Summary

Complete the two v4.0.0 operational edge-case contracts under #355:

- Field initializers: document that a `Random.Range` call in an initializer is evaluated by the Editor and stored as a baked default, not runtime randomness. `Start()` / lazy-init is reserved for local or per-client randomness. Shared per-object or per-session seed/state is generated by the owner, stored in `[UdonSynced]`, and sent after ownership is established with `RequestSerialization()` in Manual sync; receivers apply derived state in `OnDeserialization()` when needed, without making that callback a prerequisite for field synchronization. Late joiners receive the current synced state.
- Synced-array revision guards: document that a late joiner's first `OnDeserialization()` receives the current revision. Durable `ApplyValues()` runs first; the shared owner/receiver `ApplyOneShotIfNeeded()` helper baselines that first revision without historical one-shot effects, and only later revisions trigger the effect. Revision remains neither ordering nor stale-packet protection.
- Late-joiner troubleshooting: state that current synced values arrive automatically, derived state is applied from `OnDeserialization()`, and joining alone is not a reason to call `RequestSerialization()`; the join-triggered resend example was removed without changing other explicit resend guidance.
- Full late-joiner census: active events/networking/constraints references and the StateMachine/SyncedObject templates no longer teach unchanged-state join resends. Legitimate join-time assignment mutations in the API object-pool example and MasterManagedPlayerPool remain intact.

The always-loaded Skill, rules, detailed references, troubleshooting, cheatsheet, and existing documentation contracts are kept aligned. #338 and #344 behavior is preserved.

## Scope

Changed only:

- `skills/unity-vrc-udon-sharp/SKILL.md`
- `skills/unity-vrc-udon-sharp/CHEATSHEET.md`
- `skills/unity-vrc-udon-sharp/rules/udonsharp-constraints.md`
- `skills/unity-vrc-udon-sharp/rules/udonsharp-networking.md`
- `skills/unity-vrc-udon-sharp/references/constraints.md`
- `skills/unity-vrc-udon-sharp/references/events.md`
- `skills/unity-vrc-udon-sharp/references/networking.md`
- `skills/unity-vrc-udon-sharp/references/troubleshooting.md`
- `skills/unity-vrc-udon-sharp/assets/templates/StateMachine.cs`
- `skills/unity-vrc-udon-sharp/assets/templates/SyncedObject.cs`
- `tests/docs/field-initializer-context-reference.test.sh`
- `tests/docs/synced-array-callback-reference.test.sh`

No README, CHANGELOG, SDK support contract, validator, unrelated networking/initializer guidance, or release metadata was changed.

## TDD evidence

- RED on base `43a10bb513248134ebaa938fc6062b40df685fa1`: the original field-initializer and synced-array contracts failed before implementation.
- RED for the follow-up shared-state context contract was captured after adding the contextual test helper and before the documentation wording change:
  - `/home/natsuki/workspaces/evidence/agent-skills-vrc-udon/issue-355/RED-shared-state-context.log`
- RED for the late-joiner correction was captured after adding the section-scoped regression contract and before the documentation edit:
  - `/home/natsuki/workspaces/evidence/agent-skills-vrc-udon/issue-355/RED-late-joiner-resend.log`
- RED for the full late-joiner resend census was captured against `8a4d042` after adding the explicit section/file-scoped contracts and before the documentation/template edits:
  - `/home/natsuki/workspaces/evidence/agent-skills-vrc-udon/issue-355/RED-late-joiner-resend-census.log`
- RED logs for the original contracts:
  - `/home/natsuki/workspaces/evidence/agent-skills-vrc-udon/issue-355/RED-field-initializer.log`
  - `/home/natsuki/workspaces/evidence/agent-skills-vrc-udon/issue-355/RED-synced-array.log`
- GREEN after `0316cd0`, `6bd35f6`, `8a4d042`, and follow-up `4a887ef`: targeted contracts pass.
- Structural coverage confirms durable `ApplyValues()` precedes the revision helper, the first revision baselines with no one-shot, and `PlayOneShot()` remains inside the shared helper.

## Validation

- All `tests/docs/*.test.sh`: PASS
- Targeted initializer, synced-array, and full late-joiner census contracts: PASS
- Positive join-time mutation contracts for the API object pool and MasterManagedPlayerPool: PASS
- Bash hook suite: 152 passed, 0 failed
- BusyBox awk portability: PASS
- PowerShell: unavailable locally; not claimed as locally run
- System skill-creator `quick_validate.py`: all three Skills report `Skill is valid!`
- `npm pack --dry-run`: PASS; required package paths present
- `git diff --check`: PASS
- Symlink integrity: PASS
- The new `RevisionGuardedArray` example and changed C# templates were checked by the Markdown structural contract and hook gates; no standalone Unity/SDK compile probe was run in this environment.

## Checklist

- [x] Base is `dev` at `43a10bb513248134ebaa938fc6062b40df685fa1`
- [x] Assignee and labels set
- [x] Scope self-reviewed; no #338/#344 weakening found
- [x] Required CI 7/7 green for `4a887ef`
- [x] CodeRabbit findings answered; all 3 review threads resolved
- [x] Ready for review

Closes #355
Refs #338, #344, #349

### CODE REVIEW

File: tests/docs/field-initializer-context-reference.test.sh [hunk 1]
```diff
@@ -30,9 +30,17 @@ CHEATSHEET="skills/unity-vrc-udon-sharp/CHEATSHEET.md"
...
+assert_contains "$SKILL" "baked default"
+assert_contains "$SKILL" "per-instance, per-client, or per-session randomness"
+assert_contains "$SKILL" "lazy-init guard"
 assert_contains "$RULES" "## Blocked in Udon Runtime"
 assert_contains "$RULES" "Editor-Evaluated Field Initializers"
+assert_contains "$RULES" 'Nondeterministic calls such as `Random.Range`'
+assert_contains "$RULES" "compiled Udon program's baked default"
 assert_contains "$REFERENCE" "### Editor-Evaluated Field Initializers vs. Udon Runtime"
+assert_contains "$REFERENCE" "Random.Range"
+assert_contains "$REFERENCE" "per-instance, per-client, or per-session randomness"
```
[the remark this answers is not in the document]
Comment (niaka3dayo) [reply tests/docs/field-initializer-context-reference.test.sh]: Implemented in 6bd35f6. The five user-facing surfaces now distinguish local/per-client randomness (Start()/lazy-init) from shared per-object/per-session state (owner-generated [UdonSynced] seed/state, ownership before RequestSerialization() for Manual sync, and receiver-side derived-state application in OnDeserialization when needed; the callback is not required for field synchronization itself). The initializer contract test now verifies these relationships within the same documentation block. Targeted and full documentation tests pass.

File: tests/docs/synced-array-callback-reference.test.sh [hunk 2]
```diff
@@ -70,7 +76,11 @@ require_text "$SYNC_EXAMPLES" 'same length, reassigning the array, or changing i
 require_text "$TROUBLESHOOTING" '### Synced Array Callback Silent Failure'
 require_text "$TROUBLESHOOTING" 'same-length element changes, array reassignments, and array length changes'
 require_text "$TROUBLESHOOTING" 'ApplyValues()'
+require_text "$TROUBLESHOOTING" 'late joiner'
+require_text "$TROUBLESHOOTING" 'baseline'
 require_text "$CHEATSHEET" 'Array contents changed: use `OnDeserialization()`'
```
[the remark this answers is not in the document]
Comment (niaka3dayo) [reply tests/docs/synced-array-callback-reference.test.sh:86]: I verified this assertion against the current source and targeted test: the single-quoted shell literal contains Markdown backticks without backslashes, so the needle matches the documented text literally and the test passes. I left the assertion unchanged.

File: skills/unity-vrc-udon-sharp/references/troubleshooting.md [hunk 3]
```diff
@@ -1651,6 +1654,14 @@ void Start()
 
 Use an initializer for pure initial value generation when the final field type and value are supported by Udon. LINQ, lambdas, and a same-behaviour static helper using `List<T>` are allowed only in that Editor-side evaluation. If the value depends on `Networking.LocalPlayer`, a scene reference, or other runtime state, assign it in `Start()` or lazy initialization instead. Constructors and field initializers can run on a loading thread, so they must not call main-thread-only Unity APIs such as `FindObjectsByType`.
 
+For runtime randomness, do not write `private int seed = Random.Range(0, 100);` as a
+field initializer. Use `Start()` or a lazy-init guard only for local or per-client
+randomness. For shared per-object or per-session seed/state, the owner generates
+it and stores it in a `[UdonSynced]` field; with Manual sync, establish ownership
+before writing and then call `RequestSerialization()`. Receivers may apply derived
+state in `OnDeserialization()` when needed, but that callback is not required for
+the field synchronization itself, and late joiners receive the current synced state.
```
[the remark this answers is not in the document]
Comment (niaka3dayo) [reply skills/unity-vrc-udon-sharp/references/troubleshooting.md:1658]: Implemented in 8a4d042. The Late Joiner State Issues section now states that late joiners automatically receive current synced values, applies derived state from OnDeserialization(), and explicitly says not to call RequestSerialization() merely because a player joined. The OnPlayerJoined resend example was removed; other explicit resend guidance outside this section is unchanged. The section-scoped regression contract and full documentation suite pass.

[the remark this answers is not in the document]
Comment (niaka3dayo) [reply skills/unity-vrc-udon-sharp/references/troubleshooting.md:1658]: Follow-up in 4a887ef extends the same late-joiner contract across events, networking, constraints, and the StateMachine/SyncedObject templates. The positive join-time assignment mutations remain covered; the thread remains resolved.

**Comment by niaka3dayo** [comment]
Review follow-up for the current HEAD (`4a887ef5`):

- `6bd35f6` separates local/per-client randomness from owner-generated synced state and strengthens the initializer contract tests.
- `8a4d042` removes the remaining join-triggered resend guidance from the late-joiner troubleshooting section.
- `4a887ef` completes the corpus-wide pass: unchanged-state resends were removed from active references and distributed templates, while the two join-time examples that actually mutate authoritative state remain covered by structural tests.

All three review threads are resolved. A full-diff independent review of `4a887ef5` returned APPROVE (119/120) with no CRITICAL, HIGH, or MUST findings. All seven required checks and GitGuardian are green. The latest CodeRabbit pass was rate-limited, so the two `CHANGES_REQUESTED` reviews on `0316cd0` and `6bd35f6` now describe earlier revisions rather than the current head.
